File: backend/app/agent/agent_service.py
"""Service for managing AI agent definitions."""

from ampf.base import BaseFactory
from app.config import ServerConfig
from .agent_model import Agent


class AgentService:
    """Service for managing AI agent definitions."""

    def __init__(self, config: ServerConfig, factory: BaseFactory):
        self.config = config
        self.factory = factory

    def _crete_storage(self, user_email: str):
        """Create storage for agents of a given user."""
        return self.factory.create_compact_storage(
            f"users/{user_email}/agents", Agent, key_name="name"
        )

    def create_default(self, user_email: str, ai_model_name: str):
        """Create a default agent for a given model."""
        agent = Agent(
            name=ai_model_name,
            description=f"Default agent for {ai_model_name}",
            ai_model_name=ai_model_name,
            system_prompt="",
        )
        self.create(user_email, agent)
        return agent

    def create(self, user_email: str, agent: Agent):
        """Create a new agent for a given user."""
        storage = self._crete_storage(user_email)
        storage.save(agent)

    def get_all(self, user_email: str):
        """Return a list of agent names for a given user."""
        storage = self._crete_storage(user_email)
        ret = list(storage.get_all())
        if not ret:
            ai_model_name = self.config.default_model
            agent = self.create_default(user_email, ai_model_name=ai_model_name)
            ret = [agent]
        return ret

    def get(self, user_email: str, agent_name: str):
        """Return an agent by name."""
        storage = self._crete_storage(user_email)
        return storage.get(agent_name)

    def put(self, user_email: str, agent_name: str, agent: Agent):
        """Update an agent."""
        storage = self._crete_storage(user_email)
        storage.put(agent_name, agent)

    def delete(self, user_email: str, agent_name: str):
        """Delete an agent."""
        storage = self._crete_storage(user_email)
        storage.delete(agent_name)

File: frontend/src/app/agent/agent.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

export interface Agent {
    name: string;
    description: string;
    ai_model_name: string;
    system_prompt: string;
    keywords: string[];
}

@Injectable({
    providedIn: 'root'
})
export class AgentService {
    
    private apiUrl = '/api/agents';

    constructor(private http: HttpClient) { }

    get_all(): Observable<Agent[]> {
        return this.http.get<Agent[]>(this.apiUrl);
    }

    get(agentName: string): Observable<Agent> {
        return this.http.get<Agent>(`${this.apiUrl}/${agentName}`);
    }

    create(agent: Agent): Observable<Agent> {
        return this.http.post<Agent>(this.apiUrl, agent);
    }

    update(agentName: string, agent: Agent): Observable<Agent> {
        return this.http.put<Agent>(`${this.apiUrl}/${agentName}`, agent);
    }

    delete(agentName: string): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${agentName}`);
    }

}

